### PR TITLE
[don't merge] IRGen should emit code for unused private functions at -Onone -g

### PR DESCRIPTION
**Pull request by swiftix**
#### Resolved bug number: ([SR-1667](https://bugs.swift.org/browse/SR-1667))

rdar://26625045

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

This allows for setting breakpoints on such functions and calling them when evaluating expressions in the debugger.

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci Please test

File: lib/IRGen/GenDecl.cpp [hunk 1]
```diff
@@ -802,8 +802,19 @@ void IRGenerator::emitGlobalTopLevel() {
...
-    if (!isPossiblyUsedExternally(f.getLinkage(), isWholeModule))
+    if (!isPossiblyUsedExternally(f.getLinkage(), isWholeModule)) {
+      // This function is not externally visible.
+      // It could be still useful to emit it for debug purposes if it is
+      // a debug build with disabled optimizations.
+      // Do not emit this function if it is coming from a binary SIL file
+      // or if debug info is not supposed to be emitted or if it is a JIT mode.
+      if (!PrimaryIGM->IRGen.Opts.Optimize && !PrimaryIGM->IRGen.Opts.UseJIT &&
+          f.hasLocation() && !f.getLocation().isSILFile() &&
+          PrimaryIGM->DebugInfo) {
```
Comment (adrian-prantl) [inline lib/IRGen/GenDecl.cpp]: In all of LLVM it is considered a bug if the debug options affect codegen. IMHO, we should also try to live by this rule in Swift. Practically, there are few reasons to build at -Onone without debug information anyway, so I think we don't loose much by just deciding based on IRGen.Opts.Optimize.

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by CodaFi** [comment]
@swiftix Is this patch still necessary?

**Comment by swiftix** [comment]
No. We can close it.

**Pull request closed without merging**